modules/barcode/src/detector/bardetect.cpp

```diff
@@ -0,0 +1,422 @@
+
+        for (int pos = 0; pos < new_size.width; pos++)
+        {
+
+            // then calculate the column locations of the rectangle and set them to -1
+            // if they are outside the matrix bounds
+            if (pos * window_size >= width)
+            {
+                continue;
+            }
+            left_col = pos * window_size;
+            right_col = min(width, (pos + 1) * window_size);
+
+            //we had an integral image to count non-zero elements
+            CALCULATE_SUM(edges_ptr, rect_area)
+            if (rect_area < THRESHOLD_AREA)
+            {
+                // smooth region
+                coherence_row[pos] = 0;
+                continue;
+            }
+
+            CALCULATE_SUM(x_sq_ptr, x_sq)
+            CALCULATE_SUM(y_sq_ptr, y_sq)
+            CALCULATE_SUM(xy_ptr, xy)
+
+            // get the values of the rectangle corners from the integral image - 0 if outside bounds
+            d = sqrt((x_sq - y_sq) * (x_sq - y_sq) + 4 * xy * xy) / (x_sq + y_sq);
+            if (d > THRESHOLD_COHERENCE)
+            {
+                coherence_row[pos] = 255;
+                orientation_row[pos] = computeOrientation(x_sq - y_sq, 2 * xy) / 2.0f;
+                edge_nums_row[pos] = rect_area;
+            }
+            else
+            {
+                coherence_row[pos] = 0;
+            }
+
+        }
+
+    }
+}
+
+// will change localization_bbox bbox_scores
+// will change coherence,
+// depend on coherence orientation edge_nums
+void Detect::regionGrowing(int window_size)
+{
+    static constexpr float LOCAL_THRESHOLD_COHERENCE = 0.95f, THRESHOLD_RADIAN =
+            PI / 30, THRESHOLD_BLOCK_NUM = 35, LOCAL_RATIO = 0.5f, EXPANSION_FACTOR = 1.2f;
+    Point pt_to_grow, pt;                       //point to grow
+
+    float src_value;
+    float cur_value;
+    float edge_num;
+    float rect_orientation;
+    float sin_sum, cos_sum, counter;
+    //grow direction
+    static constexpr int DIR[8][2] = {{-1, -1},
+                                      {0,  -1},
+                                      {1,  -1},
+                                      {1,  0},
+                                      {1,  1},
+                                      {0,  1},
+                                      {-1, 1},
+                                      {-1, 0}};
+    vector<Point2f> growingPoints, growingImgPoints;
+    for (int y = 0; y < coherence.rows; y++)
+    {
+        auto *coherence_row = coherence.ptr<uint8_t>(y);
+
+        for (int x = 0; x < coherence.cols; x++)
+        {
+            if (coherence_row[x] == 0)
+            {
+                continue;
+            }
+            // flag
+            coherence_row[x] = 0;
+            growingPoints.clear();
+            growingImgPoints.clear();
+
+            pt = Point(x, y);
+            cur_value = orientation.at<float_t>(pt);
+            sin_sum = sin(2 * cur_value);
+            cos_sum = cos(2 * cur_value);
```

Comment:
BTW, in general `sin`/`cos` computation in nested loops is not very cheap.

Reply:
Do you have some suggestion for this? I analyzed the performance by Visual Studio, found that 
![image](https://user-images.githubusercontent.com/43094337/114976603-d6368380-9eb8-11eb-815a-6f02dfe16ed0.png)
It seems like preprocess() is much more time-consuming than this.

Reply:
Thank you for checking this.
You are right, this is not critical part of computations for now.

Possible optimization are usually done through using of pre-computed sin/cos tables (with input value quantization).